golem-component-service-base/src/repo/component.rs

```diff
@@ -0,0 +1,467 @@
+
+    async fn get_latest_version(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+    ) -> Result<Option<ComponentRecord>, RepoError>;
+
+    async fn get_by_version(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+        version: u64,
+    ) -> Result<Option<ComponentRecord>, RepoError>;
+
+    async fn get_by_name(
+        &self,
+        namespace: &str,
+        name: &str,
+    ) -> Result<Vec<ComponentRecord>, RepoError>;
+
+    async fn get_ids_by_name(&self, namespace: &str, name: &str) -> Result<Vec<Uuid>, RepoError>;
+
+    async fn get_namespaces(&self, component_id: &Uuid) -> Result<Vec<(String, i64)>, RepoError>;
+
+    async fn delete(&self, namespace: &str, component_id: &Uuid) -> Result<(), RepoError>;
+}
+
+pub struct DbComponentRepo<DB: Database> {
+    db_pool: Arc<Pool<DB>>,
+}
+
+impl<DB: Database> DbComponentRepo<DB> {
+    pub fn new(db_pool: Arc<Pool<DB>>) -> Self {
+        Self { db_pool }
+    }
+}
+
+#[async_trait]
+impl ComponentRepo for DbComponentRepo<sqlx::Sqlite> {
+    async fn upsert(&self, component: &ComponentRecord) -> Result<(), RepoError> {
+        let mut transaction = self.db_pool.begin().await?;
+
+        let result = sqlx::query(
+            "SELECT count(component_id) as count FROM components WHERE namespace != $1 AND component_id = $2",
+        )
+        .bind(component.namespace.clone())
+        .bind(component.component_id)
+        .fetch_one(&mut *transaction)
+        .await?;
+
+        let count: i64 = result.get("count");
+
+        if count > 0 {
+            return Err(RepoError::Internal("Component id not unique".to_string()));
+        }
+
+        sqlx::query(
+            r#"
+              INSERT INTO components
+                (namespace, component_id, version, name, size, user_component, protected_component, protector_version, metadata)
+              VALUES
+                ($1, $2, $3, $4, $5, $6, $7, $8, $9)
+              ON CONFLICT (namespace, component_id, version) DO UPDATE
+              SET name = $4,
+                  size = $5,
+                  user_component = $6,
+                  protected_component = $7,
+                  protector_version = $8,
+                  metadata = $9
+               "#,
+        )
+            .bind(component.namespace.clone())
+            .bind(component.component_id)
+            .bind(component.version)
+            .bind(component.name.clone())
+            .bind(component.size)
+            .bind(component.user_component.clone())
+            .bind(component.protected_component.clone())
+            .bind(component.protector_version)
+            .bind(component.metadata.clone())
+            .execute(&mut *transaction)
+            .await?;
+
+        transaction.commit().await?;
+
+        Ok(())
+    }
+
+    async fn get(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
```

Comment:
If component id should be unique - why would we need a namespace for `get`?

Reply:
I think data model should reflect access concerns. I kind of get your question though and its valid, but my opinion here is , I think Peter is right here with having namespace and it clearly reflects what's going on with the ownership and access.

Reply:
Yes, ownership is quite important, but I believe we should keep all the authorization-related code in the service.
We might not even know the namespace on `get` - with the planned `URN` schema it should be `component_id` only. User can access component by id in other user account if he has such permissions. And the user should not provide project id for this.